org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/ActionImpl.java

```diff
@@ -449,76 +420,307 @@ public ByteData collect(String inputName, InputStream inputStream, int inputCoun
-		return FileUtils.reader(buffer);
+	/**
+	 * Copy a file to a folder. Use the input file relative path as the output
+	 * file relative path.
+	 *
+	 * @param inputPath The relative path of the input file.
+	 * @param inputFile The file which is to be copied.
+	 * @param outputFolder The folder into which to copy the file.
+	 * @throws TransformException Thrown if the copy fails.
+	 */
+	protected void copyInto(String inputPath, File inputFile, File outputFolder) throws TransformException {
+		copyInto(inputPath, inputFile, outputFolder, inputPath);
+	}
+
+	/**
+	 * Copy a file to a folder.
+	 *
+	 * @param inputPath The relative path of the input file.
+	 * @param inputFile The file which is to be copied.
+	 * @param outputFolder The folder into which to copy the file.
+	 * @param outputPath The relative path of the output file.
+	 * @throws TransformException Thrown if the copy fails.
+	 */
+	protected void copyInto(String inputPath, File inputFile, File outputFolder, String outputPath)
+		throws TransformException {
+		File outputFile;
+		try {
+			outputFile = IO.getBasedFile(outputFolder, outputPath);
+		} catch (IOException e) {
+			throw new TransformException(
+				"Non-valid file [ " + outputFolder.getAbsolutePath() + " ] [ " + outputPath + " ]", e);
+		}
+
+		File parentFile = outputFile.getParentFile();
+		try {
+			IO.mkdirs(parentFile);
+		} catch (IOException e) {
+			throw new TransformException(
+				"Failed to create parent directory of [ " + outputFile.getAbsolutePath() + " ]", e);
+		}
+
+		try {
+			IO.copy(inputFile, outputFile);
+		} catch (IOException e) {
+			throw new TransformException(
+				"Failed to copy [ " + inputFile.getAbsolutePath() + " ] to [ " + outputFile.getAbsolutePath() + " ]",
+				e);
+		}
 	}
 
-	protected BufferedWriter writer(OutputStream outputStream) {
-		return FileUtils.writer(outputStream);
+	//
+
+	@FunctionalInterface
+	public interface StringReplacement {
+		String apply(String inputName, String initialValue, List<String> cases);
 	}
 
-	@Override
-	public String relocateResource(String inputPath) {
-		String prefix;
-		String inputName;
-		if (inputPath.startsWith("WEB-INF/classes/")) {
-			prefix = "WEB-INF/classes/";
-			inputName = inputPath.substring(prefix.length());
-		} else if (inputPath.startsWith("META-INF/versions/")) {
-			prefix = "META-INF/versions/";
-			int nextSlash = inputPath.indexOf('/', prefix.length());
-			if (nextSlash != -1) {
-				prefix = inputPath.substring(0, nextSlash + 1);
+	public static final List<StringReplacement> NO_ACTIVE_REPLACEMENTS = null;
+
+	public static List<StringReplacement> asReplacements(StringReplacement... activeReplacements) {
+		if (activeReplacements.length == 0) {
+			return NO_ACTIVE_REPLACEMENTS;
+		} else {
+			List<StringReplacement> replacements = new ArrayList<StringReplacement>(activeReplacements.length);
+			for (StringReplacement replacement : activeReplacements) {
+				replacements.add(replacement);
 			}
```

Comment:
'activeReplacements' is a StringReplacement[], since it was provided via a ... parameter.  'addAll' on List doesn't support addAll with an array parameter.
But ... ' Arrays.asList(activeReplacements)' works just fine.  Fixed.

Reply:
I did not say `List`. I said `Collections.addAll` which is a static method that takes an array argument.